### ISSUE TITLE
Support emoji ZWJ sequences

### ISSUE THREAD

**Issue by xster** [flutter/flutter#9652]
iOS texts can't render zero width joiner sequences like emoji skin/gender modifiers and region indicator sequences.

Extracted from #8775

**Comment by KevinTheGray** [comment]
Do we have an estimate on this?  I'd like to build a feature the revolves around emojis and want to know what my options are here

**Comment by xster** [comment]
@jason-simmons might be able to answer. I assume this may be fixed with libtxt.

**Comment by dvoiss** [comment]
@xster I'm in the same position as @KevinTheGray.

I've found Android to be okay, but iOS can't render anything.

Could you (or @jason-simmons) tell me more about `libtxt` and if that fixes it, how I can get it into my current build?

**Comment by KevinTheGray** [comment]
to also chime in with @dvoiss 

I actually am seeing issues on Android.  for example, on a Nexus 5x running android 7.0.1, a man with skin color emoji will render the man+square_skin_color, just like on iOS

**Comment by jason-simmons** [comment]
libtxt is a new implementation of Flutter's text rendering engine.  One of the goals of libtxt is offering better support for emoji.

libtxt is currently enabled on the master branch of Flutter.  We may need to roll it back temporarily if problems arise, but our intent is to deploy it permanently soon.

**Comment by dvoiss** [comment]
@jason-simmons Could you link me to anything relevant that I can use such as a PR or some such?

To echo @KevinTheGray there are issues on Android too. Sample code and a screenshot are below.

In the screenshot you can see how the following items render, I've pasted a screenshot of how they look in the code because pasting the code itself into this comment doesn't work because github renders them correctly 😄  As the screenshot shows man + skin-color has issues, but oddly enough render okay when paired with a flag. 

But when paired with a flag a new issue arises (I don't know if there's an issue logged for this) which is seen on the second line in the screenshot where the width of the Text widget is "five emojis" in length, presumably because it's measuring the width of all the code-points but then renders them combined correctly?

**The red is so you can see the extra space on lines 2, 5, and 7.**

![screen](https://user-images.githubusercontent.com/1313555/37436784-edc0ad3c-27a6-11e8-8607-ec2751aeae7e.png)

Code:

<img width="322" alt="screen shot 2018-03-14 at 4 38 33 pm" src="https://user-images.githubusercontent.com/1313555/37436682-5435d0c0-27a6-11e8-992c-3890165b4ac3.png">

Full code below. You can see the emoji render correctly when pasted here on github. Note that I use the emoji-one font below to resolve issues with emojis on different Android versions:

```dart
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Emoji test'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final List<String> emojiTestItems = [
    '👨🏻‍🎨🎨🎨',
    '👨🏻‍🎨🎨🇮🇹',
    '👨🏻‍🎨🎨',
    '👨🏻‍🎨',
    '👨🏻',
    '👨🏻🇮🇹',
    '🇮🇹'
  ];

  List<Widget> _getEmojiTestItems() {
    List<Widget> widgets = emojiTestItems.map((item) => new Container(
      child: new Text(item, style: new TextStyle(fontSize: 48.0, fontFamily: "Emoji One")),
      color: Colors.red)
    ).toList();

    widgets.add(new Center(child: new Container(
      child: new Text('👨🏻‍🎨🎨🇮🇹', textAlign: TextAlign.center, style: new TextStyle(fontSize: 48.0, fontFamily: "Emoji One")),
      color: Colors.red)
    ));

    return widgets;
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
      body: new Center(
        child: new Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: _getEmojiTestItems(),
        ),
      ),
    );
  }
}
```

**Comment by jason-simmons** [comment]
We are currently in the middle of the transition from the original Blink-based text renderer to Libtxt (which is based on a text handling library used within Android).

Libtxt was enabled in the master branch of Flutter in this commit (which landed yesterday):
    https://github.com/flutter/flutter/commit/5a58b36e36b8d7aace89d3950e6deb307956a6a0

If you're building a custom engine then you can select the libtxt or Blink text renderer with the settings.using_blink flag:
    https://github.com/flutter/engine/blob/master/shell/common/shell.cc#L123

The Blink-based engine is considered deprecated.  After Libtxt is fully landed we can look at improving its emoji support.

**Comment by dvoiss** [comment]
@jason-simmons Thanks for the info. 

I updated my local Flutter to the latest.

But is there any way to test libtxt without the whole rigamarole of building the engine and debugging whatever build errors will inevitably crop up?

Something like `flutter run --enable-blink false`?

**Comment by jason-simmons** [comment]
If you switch to the master branch of the Flutter framework today you'll get an engine that uses Libtxt.

We don't have a way to switch between Libtxt and Blink in the Flutter tools.  The goal is to remove Blink from the engine soon.

**Comment by dvoiss** [comment]
> If you switch to the master branch of the Flutter framework today you'll get an engine that uses Libtxt.

Interesting, so since I still see issues then things weren't just magically fixed like I was hoping 🤣 

I figured it wasn't enabled by default because I saw this and thought using blink was defaulted to true: https://github.com/flutter/engine/blob/master/common/settings.h#L28.

**Comment by rlee1990** [comment]
Any fix yet?

**Comment by jonasbark** [comment]
Unfortunately no. I had some hopes when https://github.com/flutter/flutter/issues/20303 landed but no dice. 
@xster any chance of prioritizing this alongside the Cupertino widget support efforts?

**Comment by lukepighetti** [comment]
As someone who just added chat to his app, emojis are not a nicety, but a necessity for UX. I do not prioritize one platform over the other, so if it's a blocking situation on iOS I consider it a blocking situation for both platforms because the feature won't be released.

**Comment by lukepighetti** [comment]
@xster Can we get a timeline for this?

**Comment by dvoiss** [comment]
Any updates on this, I just thought of this game after seeing this talk on "Building the Android Summit App" with Flutter.

Specifically the builders of the app mention this bug (or a related one to emoji sequences) here (link to time-stamp): https://youtu.be/xENT8qPIazY?list=PLzJZrgVJE8BbctT8_LBqUUHDzFTjKsEkU&t=1464

**Comment by adriancmurray** [comment]
Are there any updates? I'd love for this to get patched up.

**Comment by zoechi** [comment]
I guess it is blocked on https://github.com/dart-lang/sdk/issues/28404

**Comment by jason-simmons** [comment]
I took a look at this a while ago.  This bug is related to the integration between libtxt, HarfBuzz, and Skia.  IIRC, HarfBuzz is relying on CoreText to render ZWJ sequences on Apple platforms.  But libtxt/Skia is not providing the needed CoreText fonts to HarfBuzz.

libtxt's `FontSkia::CreateHarfBuzzFace` is extracting the font data from an `SkTypeface` and calling `hb_face_create_for_tables`.  On Apple platforms, perhaps ZWJ rendering would work if it instead used `SkTypeface_GetCTFontRef` and `hb_coretext_face_create` to create a HarfBuzz face based on a CoreText font?

However, Skia says that `SkTypeface_GetCTFontRef` is a deprecated API, so that isn't a sustainable solution.

@chinmaygarde @GaryQian

**Comment by khaledhosny** [comment]
Recently released HarfBuzz 2.0.0 has native support for AAT fonts, so it should be able to handle ZWJ sequences for Apple Color Emoji without Core Text integration.

**Comment by GaryQian** [comment]
Will begin with an update to harfbuzz 2.1.0. If all goes well, this might be fixed by magic! If not, I can take a deeper look at it.
https://github.com/flutter/flutter/pull/24024
https://github.com/flutter/engine/pull/6760

**Comment by jonasbark** [comment]
Just checked out newest master and yeah, finally there they are.

![image](https://user-images.githubusercontent.com/1151304/48117296-08b01100-e269-11e8-810f-177d9e58cc5a.png)

**Comment by GaryQian** [comment]
Have confirmed zwj support on iOS. Feel free to reopen/comment here is there are any issues or more work to be done!

**Comment by adriancmurray** [comment]
@GaryQian I'm running on 0.10.2 and while there is an improvement, I think there still might be some work to be done. 
<img width="61" alt="screen shot 2018-11-11 at 12 42 48 pm" src="https://user-images.githubusercontent.com/39349880/48317430-6026e780-e5af-11e8-834c-55aa0170dcaf.png">
This the David Bowie (or rock star...) emoji as an example. There are a lot of emjois that are still not working properly on iOS. Or maybe this is a different issue....

**Comment by skabber** [comment]
> @GaryQian I'm running on 0.10.2 and while there is an improvement, I think there still might be some work to be done.
> <img alt="screen shot 2018-11-11 at 12 42 48 pm" width="61" src="https://user-images.githubusercontent.com/39349880/48317430-6026e780-e5af-11e8-834c-55aa0170dcaf.png">
> This the David Bowie (or rock star...) emoji as an example. There are a lot of emjois that are still not working properly on iOS. Or maybe this is a different issue....

This look correct to me in 0.11.7.

**Comment by knopp** [comment]
The rendering seems to work, but cursor movement seems broken (tested on macOS). @GaryQian, any ideas?
